dji_sdk/src/modules/dji_sdk_node_main.cpp

```diff
@@ -39,11 +51,27 @@ void DJISDKNode::broadcast_callback()
    
     auto current_time = ros::Time::now();
 
+
+
     if(msg_flags & HAS_TIME){
+
+        // Only this message containes the received time in the header and the hardware time within the message
+        // All following messages should have translated hardware times in the header.
         time_stamp.header.frame_id = "/time";
         time_stamp.header.stamp = current_time;
-        time_stamp.time = bc_data.timeStamp.time;
-        time_stamp.time_ns = bc_data.timeStamp.nanoTime;
+
+        if(device_time_translator_){
+        current_time = device_time_translator_->update(bc_data.timeStamp.time, current_time, 0.0); // estimated offset is 0.0 for now observed to usuall be < 2 ms can be upto ~ 20 ms!!!
+        //cout << setprecision(15) <<setw(30) << "Translated HW Time Offset: " << current_time.toSec() - ros::Time::now().toSec() << endl;
+        }
+        time_stamp.time = bc_data.timeStamp.time;  // This is now in units of 1/400 s
+        time_stamp.time_ns = bc_data.timeStamp.nanoTime; // This is perhaps more accurate??? But requires handling the wrapping after every 4.295 secs. DO NOT SUM both as it leads to double counting.
+        if(bc_data.timeStamp.nanoTime < ::g_prevNanoTime){
+            ::g_counter++;
+        }
+        g_hardwareStamp = ::g_counter*::g_WRAP_TIME + double(bc_data.timeStamp.nanoTime)*(1e-9);
```

Comment:
The cuckoo_time_translator packages provides a class that does this for you. It is also used within the update call up in line 64. If the nanoTime proves to be a proper wrapping hardware clock you can just use it instead of bc_data.timeStamp.time in line 64 - of course you would need to update the information given to the constructor of the device_time_translator.